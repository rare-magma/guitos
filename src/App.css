body {
  background: var(--bgcolor);
  color: var(--textcolor);
}

a {
  color: var(--selection);
  text-decoration: none;
}

a:hover {
  border-bottom: 1px solid;
  color: var(--cyan);
}

.btn:focus-visible {
  background-color: var(--comment);
  box-shadow: none !important;
  color: var(--lightbgcolor);
}

.btn.show,
.btn:hover {
  background-color: var(--comment);
  color: var(--lightbgcolor);
}

.btn-close,
.btn-currency,
.btn-outline-secondary {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
}

.btn-outline-info {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--cyan);
}

.btn-outline-info:focus-visible,
.btn-outline-info:hover {
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-outline-primary {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--purple);
}

.btn-outline-primary:focus-visible,
.btn-outline-primary:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-outline-success {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--highlight);
}

.btn-outline-success:focus-visible,
.btn-outline-success:hover {
  background-color: var(--highlight);
  border: 1px solid var(--highlight);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.card {
  border: 1px solid var(--comment);
  background: var(--lightbgcolor);
  color: var(--textcolor);
}

.card-header {
  border-bottom: 1px solid var(--comment);
}

.card-footer {
  border-top: 1px solid var(--comment);
}

.fixed-width-font,
.textarea {
  font-family:
    ui-monospace,
    SF Mono,
    Menlo,
    Monaco,
    Andale Mono,
    monospace;
}

.form-control,
.input-group-text,
input .input-group-text,
input.form-control,
input.text-end.form-control {
  background: var(--lightbgcolor);
  border: 1px solid var(--comment);
  color: var(--textcolor);
}

.form-control:focus,
input.text-end.form-control:focus {
  background: var(--lightbgcolor);
  border: 1px solid var(--pink);
  box-shadow: 0 0 0;
  color: var(--textcolor);
}

/* Chrome, Safari, Edge, Opera */
input::-webkit-inner-spin-button,
input::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

/* Firefox */
input[type="number"] {
  -moz-appearance: textfield;
}

.input-group-sm > .form-control {
  font-size: 1rem;
}

.popover {
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  width: auto;
}

.popover,
.popover-header {
  background: var(--lightbgcolor);
  color: var(--textcolor);
}

.popover-header {
  border-bottom: 0 solid var(--comment);
  border: solid var(--comment);
  border-width: 0 0 1px;
}

.popover .popover-arrow:after,
.popover .popover-arrow:before {
  border-bottom-color: var(--comment);
  border-top-color: var(--comment);
}

.pre,
.textarea {
  background: var(--selection);
  border-radius: 0;
  font-size: 0.9em;
  margin: 0.8em 0 1em;
  max-height: 35vh;
  overflow: auto;
  padding: 0.6em 0.9em;
}

.straight-corners {
  border-radius: 0;
}
