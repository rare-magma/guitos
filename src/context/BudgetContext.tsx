import { PropsWithChildren, createContext, useContext, useState } from "react";
import useUndo from "use-undo";
import { Budget } from "../components/Budget/Budget";
import { SearchOption } from "../components/NavBar/NavBar";
import { calcPercentage } from "../utils";
import { useGeneralContext } from "./GeneralContext";

interface BudgetContextInterface {
  budget: Budget | undefined;
  setBudget: (value: Budget | undefined, saveInHistory: boolean) => void;
  budgetList: Budget[] | undefined;
  setBudgetList: (value: Budget[] | undefined) => void;
  budgetNameList: SearchOption[] | undefined;
  setBudgetNameList: (value: SearchOption[] | undefined) => void;
  revenuePercentage: number;
  past: (Budget | undefined)[];
  future: (Budget | undefined)[];
  undo: () => void;
  redo: () => void;
  canUndo: boolean;
  canRedo: boolean;
}

const BudgetContext = createContext<BudgetContextInterface>({
  budget: undefined,
  setBudget: (value: Budget | undefined, saveInHistory: boolean) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    value;
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    saveInHistory;
  },
  budgetList: [],
  setBudgetList: (value: Budget[] | undefined) => value,
  budgetNameList: [],
  setBudgetNameList: (value: SearchOption[] | undefined) => value,
  revenuePercentage: 0,
  past: [undefined],
  future: [undefined],
  undo: () => {
    // undo
  },
  redo: () => {
    // redo
  },
  canUndo: false,
  canRedo: false,
});

function useBudget() {
  const {
    budget,
    setBudget,
    budgetList,
    setBudgetList,
    budgetNameList,
    setBudgetNameList,
    revenuePercentage,
    past,
    future,
    undo,
    redo,
    canUndo,
    canRedo,
  } = useContext(BudgetContext);

  return {
    budget,
    setBudget,
    budgetList,
    setBudgetList,
    budgetNameList,
    setBudgetNameList,
    revenuePercentage,
    past,
    future,
    undo,
    redo,
    canUndo,
    canRedo,
  };
}

function BudgetProvider({ children }: PropsWithChildren) {
  const [budgetList, setBudgetList] = useState<Budget[] | undefined>([]);
  const [budgetNameList, setBudgetNameList] = useState<
    SearchOption[] | undefined
  >([]);
  const { setNeedReload } = useGeneralContext();

  const [
    budgetState,
    {
      set: setBudget,
      undo,
      redo,
      canUndo: undoPossible,
      canRedo: redoPossible,
    },
  ] = useUndo<Budget | undefined>(undefined, { useCheckpoints: true });

  const { present: budget, past: pastState, future: futureState } = budgetState;
  const past = pastState.filter((b) => b?.id === budget?.id);
  const future = futureState.filter((b) => b?.id === budget?.id);

  const revenuePercentage = calcPercentage(
    budget?.expenses.total ?? 0,
    budget?.incomes.total ?? 0,
  );

  const canReallyUndo = undoPossible && past[past.length - 1] !== undefined;
  const canReallyRedo = redoPossible && future[0] !== undefined;

  function handleUndo() {
    if (canReallyUndo) {
      setNeedReload(true);
      undo();
      setNeedReload(false);
    }
  }

  function handleRedo() {
    if (canReallyRedo) {
      setNeedReload(true);
      redo();
      setNeedReload(false);
    }
  }

  return (
    <BudgetContext.Provider
      value={{
        budget,
        setBudget,
        budgetList,
        setBudgetList,
        budgetNameList,
        setBudgetNameList,
        revenuePercentage,
        past,
        future,
        undo: handleUndo,
        redo: handleRedo,
        canUndo: canReallyUndo,
        canRedo: canReallyRedo,
      }}
    >
      {children}
    </BudgetContext.Provider>
  );
}

// eslint-disable-next-line react-refresh/only-export-components
export { BudgetProvider, useBudget };
