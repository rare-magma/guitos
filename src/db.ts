import localforage from "localforage";

export const budgetsDB = localforage.createInstance({
  name: "guitos",
  storeName: "budgets",
});

export const optionsDB = localforage.createInstance({
  name: "guitos",
  storeName: "options",
});
