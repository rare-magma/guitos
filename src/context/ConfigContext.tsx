import { PropsWithChildren, createContext, useContext, useState } from "react";
import {
  CurrencyInputProps,
  IntlConfig,
} from "react-currency-input-field/dist/components/CurrencyInputProps";
import { optionsDB } from "../db";
import { initialCurrencyCode, userLang } from "../utils";

interface ConfigContextInterface {
  intlConfig: IntlConfig | undefined;
  setIntlConfig: (value: IntlConfig) => void;
  currency: string;
  setCurrency: (value: string) => void;
}

const ConfigContext = createContext<ConfigContextInterface>({
  intlConfig: { locale: userLang, currency: initialCurrencyCode },
  setIntlConfig: (value: IntlConfig) => value,
  currency: initialCurrencyCode,
  setCurrency: (value: string) => value,
});

function useConfig() {
  const { currency, setCurrency, intlConfig, setIntlConfig } =
    useContext(ConfigContext);

  function handleCurrency(c: string) {
    optionsDB.setItem("currencyCode", c).catch((e) => {
      throw new Error(e as string);
    });
    setCurrency(c);
    setIntlConfig({ locale: userLang, currency: c });
  }

  return { intlConfig, setIntlConfig, currency, handleCurrency };
}

function ConfigProvider({ children }: PropsWithChildren) {
  const [currency, setCurrency] = useState<string>(initialCurrencyCode);
  const [intlConfig, setIntlConfig] = useState<
    CurrencyInputProps["intlConfig"]
  >({
    locale: userLang,
    currency: currency,
  });

  return (
    <ConfigContext.Provider
      value={{ intlConfig, setIntlConfig, currency, setCurrency }}
    >
      {children}
    </ConfigContext.Provider>
  );
}

// eslint-disable-next-line react-refresh/only-export-components
export { ConfigProvider, useConfig };
