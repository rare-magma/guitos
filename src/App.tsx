import "bootstrap/dist/css/bootstrap.min.css";
import { Route, BrowserRouter as Router, Routes } from "react-router-dom";
import "./App.css";
import "./colors.css";
import BudgetPage from "./components/Budget/BudgetPage";
import { BudgetProvider } from "./context/BudgetContext";
import { ConfigProvider } from "./context/ConfigContext";

export default function App() {
  return (
    <ConfigProvider>
      <BudgetProvider>
        <Router>
          <Routes>
            <Route path="/" element={<BudgetPage />} />
            <Route path="/:name" element={<BudgetPage />} />
          </Routes>
        </Router>
      </BudgetProvider>
    </ConfigProvider>
  );
}
