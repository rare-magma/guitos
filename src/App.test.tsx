import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import { budgetsDB, optionsDB } from "./context/db";
import { budgetContextSpy, testEmptyBudgetContext } from "./setupTests";

describe("App", () => {
  const comp = <App />;

  beforeEach(() => {
    render(comp);
  });

  it("renders initial state", () => {
    cleanup();
    budgetContextSpy.mockReturnValue(testEmptyBudgetContext);
    render(comp);
    expect(screen.getAllByText("guitos")[0]).toBeInTheDocument();
    expect(screen.getByRole("status")).toBeInTheDocument();
    expect(screen.getByLabelText("open guitos changelog")).toBeInTheDocument();
    expect(budgetsDB.config("name")).toBe("guitos");
    expect(budgetsDB.config("storeName")).toBe("budgets");
    expect(optionsDB.config("name")).toBe("guitos");
    expect(optionsDB.config("storeName")).toBe("options");
  });

  it("shows new budget when clicking new button", async () => {
    const newButton = screen.getAllByRole("button", { name: "new budget" });
    await userEvent.click(newButton[0]);
    expect(screen.getByLabelText("delete budget")).toBeInTheDocument();
    expect(screen.getByLabelText("clone budget")).toBeInTheDocument();
    expect(screen.getByText("Statistics")).toBeInTheDocument();
    expect(screen.getByText("Revenue")).toBeInTheDocument();
    expect(screen.getByText("Expenses")).toBeInTheDocument();
  });

  it.skip("deletes budget when clicking delete button", async () => {
    const newButton = screen.getAllByRole("button", { name: "new budget" });
    await userEvent.click(newButton[0]);
    await screen
      .findAllByRole("button", {
        name: "delete budget",
      })
      .then((e) => userEvent.click(e[0]));

    await userEvent.click(
      screen.getByRole("button", { name: "confirm budget deletion" }),
    );

    expect(screen.queryByLabelText("delete budget")).not.toBeInTheDocument();
    expect(screen.queryByLabelText("clone budget")).not.toBeInTheDocument();
    expect(screen.queryByText("Statistics")).not.toBeInTheDocument();
    expect(screen.queryByText("Revenue")).not.toBeInTheDocument();
    expect(screen.queryByText("Expenses")).not.toBeInTheDocument();
  });
});
