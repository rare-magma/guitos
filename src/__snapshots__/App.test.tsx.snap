// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
      >
        <nav
          class="navbar navbar-expand-md navbar-light"
          data-testid="header"
        >
          <div
            class="flex-row container-fluid"
          >
            <span
              class="flex-sm-row navbar-brand"
            >
              <a
                aria-label="open guitos repository"
                class="brand"
                href="https://github.com/rare-magma/guitos"
                rel="noreferrer"
                target="_blank"
              >
                guitos
              </a>
            </span>
            <div
              class="flex-row flex-grow-1 navbar-nav"
            >
              <div
                class="flex-grow-1 navbar-nav"
              >
                <button
                  aria-controls="offcanvasNavbar-expand-md"
                  aria-label="Toggle navigation"
                  class="ms-auto ms-2 my-2 navbar-toggler collapsed"
                  type="button"
                >
                  <span
                    class="navbar-toggler-icon"
                  />
                </button>
              </div>
            </div>
            <div
              aria-labelledby="offcanvasNavbarLabel-expand-md"
              class="offcanvas offcanvas-end"
              id="offcanvasNavbar-expand-md"
            >
              <div
                class="offcanvas-header"
              >
                <div
                  class="offcanvas-title h5"
                  id="offcanvasNavbarLabel-expand-md"
                >
                  guitos
                </div>
                <button
                  aria-label="close navigation"
                  class="btn btn-outline-secondary"
                  type="button"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 16 16"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M2.146 2.854a.5.5 0 1 1 .708-.708L8 7.293l5.146-5.147a.5.5 0 0 1 .708.708L8.707 8l5.147 5.146a.5.5 0 0 1-.708.708L8 8.707l-5.146 5.147a.5.5 0 0 1-.708-.708L7.293 8 2.146 2.854Z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="justify-content-end offcanvas-body"
              >
                <div
                  class="navbar-nav"
                >
                  <div
                    class="m-2 navbar-nav"
                  />
                </div>
                <div
                  class="navbar-nav"
                >
                  <div
                    class="m-2 nav-item"
                  >
                    <button
                      aria-label="new budget"
                      class="w-100 btn btn-outline-success"
                      type="button"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                  <li
                    class="m-2 navbar-nav"
                  >
                    <div>
                      <button
                        aria-label="import budget"
                        class="w-100 btn btn-outline-primary"
                        type="button"
                      >
                        <svg
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 16 16"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M.5 9.9a.5.5 0 0 1 .5.5v2.5a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-2.5a.5.5 0 0 1 1 0v2.5a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2v-2.5a.5.5 0 0 1 .5-.5z"
                          />
                          <path
                            d="M7.646 1.146a.5.5 0 0 1 .708 0l3 3a.5.5 0 0 1-.708.708L8.5 2.707V11.5a.5.5 0 0 1-1 0V2.707L5.354 4.854a.5.5 0 1 1-.708-.708l3-3z"
                          />
                        </svg>
                      </button>
                      <input
                        class="form-control"
                        data-testid="import-form-control"
                        id="import"
                        multiple=""
                        style="display: none;"
                        type="file"
                      />
                    </div>
                  </li>
                  <div
                    class="m-2 nav-item"
                  >
                    <a
                      aria-label="open instructions in new tab"
                      class="w-100 btn btn-outline-info"
                      href="https://github.com/rare-magma/guitos#getting-started"
                      role="button"
                      tabindex="0"
                      target="_blank"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M4.475 5.458c-.284 0-.514-.237-.47-.517C4.28 3.24 5.576 2 7.825 2c2.25 0 3.767 1.36 3.767 3.215 0 1.344-.665 2.288-1.79 2.973-1.1.659-1.414 1.118-1.414 2.01v.03a.5.5 0 0 1-.5.5h-.77a.5.5 0 0 1-.5-.495l-.003-.2c-.043-1.221.477-2.001 1.645-2.712 1.03-.632 1.397-1.135 1.397-2.028 0-.979-.758-1.698-1.926-1.698-1.009 0-1.71.529-1.938 1.402-.066.254-.278.461-.54.461h-.777ZM7.496 14c.622 0 1.095-.474 1.095-1.09 0-.618-.473-1.092-1.095-1.092-.606 0-1.087.474-1.087 1.091S6.89 14 7.496 14Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </a>
                  </div>
                  <span
                    class="version justify-content-end align-self-end m-2 navbar-brand"
                  >
                    <a
                      aria-label="open guitos changelog"
                      href="https://github.com/rare-magma/guitos/blob/main/CHANGELOG.md"
                      rel="noreferrer"
                      target="_blank"
                    >
                      v
                      0.23.0
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </nav>
        <div
          class="position-absolute top-50 start-50 translate-middle container-fluid"
        >
          <div
            class="justify-content-center row"
          >
            <div
              class="spinner-border"
              role="status"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="container-fluid"
      >
        <nav
          class="navbar navbar-expand-md navbar-light"
          data-testid="header"
        >
          <div
            class="flex-row container-fluid"
          >
            <span
              class="flex-sm-row navbar-brand"
            >
              <a
                aria-label="open guitos repository"
                class="brand"
                href="https://github.com/rare-magma/guitos"
                rel="noreferrer"
                target="_blank"
              >
                guitos
              </a>
            </span>
            <div
              class="flex-row flex-grow-1 navbar-nav"
            >
              <div
                class="flex-grow-1 navbar-nav"
              >
                <button
                  aria-controls="offcanvasNavbar-expand-md"
                  aria-label="Toggle navigation"
                  class="ms-auto ms-2 my-2 navbar-toggler collapsed"
                  type="button"
                >
                  <span
                    class="navbar-toggler-icon"
                  />
                </button>
              </div>
            </div>
            <div
              aria-labelledby="offcanvasNavbarLabel-expand-md"
              class="offcanvas offcanvas-end"
              id="offcanvasNavbar-expand-md"
            >
              <div
                class="offcanvas-header"
              >
                <div
                  class="offcanvas-title h5"
                  id="offcanvasNavbarLabel-expand-md"
                >
                  guitos
                </div>
                <button
                  aria-label="close navigation"
                  class="btn btn-outline-secondary"
                  type="button"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 16 16"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M2.146 2.854a.5.5 0 1 1 .708-.708L8 7.293l5.146-5.147a.5.5 0 0 1 .708.708L8.707 8l5.147 5.146a.5.5 0 0 1-.708.708L8 8.707l-5.146 5.147a.5.5 0 0 1-.708-.708L7.293 8 2.146 2.854Z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="justify-content-end offcanvas-body"
              >
                <div
                  class="navbar-nav"
                >
                  <div
                    class="m-2 navbar-nav"
                  />
                </div>
                <div
                  class="navbar-nav"
                >
                  <div
                    class="m-2 nav-item"
                  >
                    <button
                      aria-label="new budget"
                      class="w-100 btn btn-outline-success"
                      type="button"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                  <li
                    class="m-2 navbar-nav"
                  >
                    <div>
                      <button
                        aria-label="import budget"
                        class="w-100 btn btn-outline-primary"
                        type="button"
                      >
                        <svg
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 16 16"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M.5 9.9a.5.5 0 0 1 .5.5v2.5a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-2.5a.5.5 0 0 1 1 0v2.5a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2v-2.5a.5.5 0 0 1 .5-.5z"
                          />
                          <path
                            d="M7.646 1.146a.5.5 0 0 1 .708 0l3 3a.5.5 0 0 1-.708.708L8.5 2.707V11.5a.5.5 0 0 1-1 0V2.707L5.354 4.854a.5.5 0 1 1-.708-.708l3-3z"
                          />
                        </svg>
                      </button>
                      <input
                        class="form-control"
                        data-testid="import-form-control"
                        id="import"
                        multiple=""
                        style="display: none;"
                        type="file"
                      />
                    </div>
                  </li>
                  <div
                    class="m-2 nav-item"
                  >
                    <a
                      aria-label="open instructions in new tab"
                      class="w-100 btn btn-outline-info"
                      href="https://github.com/rare-magma/guitos#getting-started"
                      role="button"
                      tabindex="0"
                      target="_blank"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M4.475 5.458c-.284 0-.514-.237-.47-.517C4.28 3.24 5.576 2 7.825 2c2.25 0 3.767 1.36 3.767 3.215 0 1.344-.665 2.288-1.79 2.973-1.1.659-1.414 1.118-1.414 2.01v.03a.5.5 0 0 1-.5.5h-.77a.5.5 0 0 1-.5-.495l-.003-.2c-.043-1.221.477-2.001 1.645-2.712 1.03-.632 1.397-1.135 1.397-2.028 0-.979-.758-1.698-1.926-1.698-1.009 0-1.71.529-1.938 1.402-.066.254-.278.461-.54.461h-.777ZM7.496 14c.622 0 1.095-.474 1.095-1.09 0-.618-.473-1.092-1.095-1.092-.606 0-1.087.474-1.087 1.091S6.89 14 7.496 14Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </a>
                  </div>
                  <span
                    class="version justify-content-end align-self-end m-2 navbar-brand"
                  >
                    <a
                      aria-label="open guitos changelog"
                      href="https://github.com/rare-magma/guitos/blob/main/CHANGELOG.md"
                      rel="noreferrer"
                      target="_blank"
                    >
                      v
                      0.23.0
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </nav>
        <div
          class="position-absolute top-50 start-50 translate-middle container-fluid"
        >
          <div
            class="justify-content-center row"
          >
            <div
              class="spinner-border"
              role="status"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
    >
      <nav
        class="navbar navbar-expand-md navbar-light"
        data-testid="header"
      >
        <div
          class="flex-row container-fluid"
        >
          <span
            class="flex-sm-row navbar-brand"
          >
            <a
              aria-label="open guitos repository"
              class="brand"
              href="https://github.com/rare-magma/guitos"
              rel="noreferrer"
              target="_blank"
            >
              guitos
            </a>
          </span>
          <div
            class="flex-row flex-grow-1 navbar-nav"
          >
            <div
              class="flex-grow-1 navbar-nav"
            >
              <button
                aria-controls="offcanvasNavbar-expand-md"
                aria-label="Toggle navigation"
                class="ms-auto ms-2 my-2 navbar-toggler collapsed"
                type="button"
              >
                <span
                  class="navbar-toggler-icon"
                />
              </button>
            </div>
          </div>
          <div
            aria-labelledby="offcanvasNavbarLabel-expand-md"
            class="offcanvas offcanvas-end"
            id="offcanvasNavbar-expand-md"
          >
            <div
              class="offcanvas-header"
            >
              <div
                class="offcanvas-title h5"
                id="offcanvasNavbarLabel-expand-md"
              >
                guitos
              </div>
              <button
                aria-label="close navigation"
                class="btn btn-outline-secondary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2.146 2.854a.5.5 0 1 1 .708-.708L8 7.293l5.146-5.147a.5.5 0 0 1 .708.708L8.707 8l5.147 5.146a.5.5 0 0 1-.708.708L8 8.707l-5.146 5.147a.5.5 0 0 1-.708-.708L7.293 8 2.146 2.854Z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="justify-content-end offcanvas-body"
            >
              <div
                class="navbar-nav"
              >
                <div
                  class="m-2 navbar-nav"
                />
              </div>
              <div
                class="navbar-nav"
              >
                <div
                  class="m-2 nav-item"
                >
                  <button
                    aria-label="new budget"
                    class="w-100 btn btn-outline-success"
                    type="button"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 16 16"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2Z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
                <li
                  class="m-2 navbar-nav"
                >
                  <div>
                    <button
                      aria-label="import budget"
                      class="w-100 btn btn-outline-primary"
                      type="button"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M.5 9.9a.5.5 0 0 1 .5.5v2.5a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-2.5a.5.5 0 0 1 1 0v2.5a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2v-2.5a.5.5 0 0 1 .5-.5z"
                        />
                        <path
                          d="M7.646 1.146a.5.5 0 0 1 .708 0l3 3a.5.5 0 0 1-.708.708L8.5 2.707V11.5a.5.5 0 0 1-1 0V2.707L5.354 4.854a.5.5 0 1 1-.708-.708l3-3z"
                        />
                      </svg>
                    </button>
                    <input
                      class="form-control"
                      data-testid="import-form-control"
                      id="import"
                      multiple=""
                      style="display: none;"
                      type="file"
                    />
                  </div>
                </li>
                <div
                  class="m-2 nav-item"
                >
                  <a
                    aria-label="open instructions in new tab"
                    class="w-100 btn btn-outline-info"
                    href="https://github.com/rare-magma/guitos#getting-started"
                    role="button"
                    tabindex="0"
                    target="_blank"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 16 16"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M4.475 5.458c-.284 0-.514-.237-.47-.517C4.28 3.24 5.576 2 7.825 2c2.25 0 3.767 1.36 3.767 3.215 0 1.344-.665 2.288-1.79 2.973-1.1.659-1.414 1.118-1.414 2.01v.03a.5.5 0 0 1-.5.5h-.77a.5.5 0 0 1-.5-.495l-.003-.2c-.043-1.221.477-2.001 1.645-2.712 1.03-.632 1.397-1.135 1.397-2.028 0-.979-.758-1.698-1.926-1.698-1.009 0-1.71.529-1.938 1.402-.066.254-.278.461-.54.461h-.777ZM7.496 14c.622 0 1.095-.474 1.095-1.09 0-.618-.473-1.092-1.095-1.092-.606 0-1.087.474-1.087 1.091S6.89 14 7.496 14Z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </a>
                </div>
                <span
                  class="version justify-content-end align-self-end m-2 navbar-brand"
                >
                  <a
                    aria-label="open guitos changelog"
                    href="https://github.com/rare-magma/guitos/blob/main/CHANGELOG.md"
                    rel="noreferrer"
                    target="_blank"
                  >
                    v
                    0.23.0
                  </a>
                </span>
              </div>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="position-absolute top-50 start-50 translate-middle container-fluid"
      >
        <div
          class="justify-content-center row"
        >
          <div
            class="spinner-border"
            role="status"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
